sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BeamFnDataReadRunnerTest.java

```diff
@@ -122,334 +129,579 @@
+                0.5,
+                0
+              })
+          // Progress into the active element influences where the split of the remainder falls.
+          .add(new Object[] {channelSplitResult(1L), 4L, ImmutableList.of("A"), 0.25, 0.5})
+          .add(new Object[] {channelSplitResult(2L), 4L, ImmutableList.of("A"), 0.25, 0.9})
+          .add(new Object[] {channelSplitResult(2L), 4L, ImmutableList.of("A", "B"), 0.25, 0.0})
+          .add(new Object[] {channelSplitResult(2L), 4L, ImmutableList.of("A", "B"), 0.25, 0.1})
+          .build();
+    }
 
-  @Test
-  public void testRegistration() {
-    for (Registrar registrar : ServiceLoader.load(Registrar.class)) {
-      if (registrar instanceof BeamFnDataReadRunner.Registrar) {
-        assertThat(
-            registrar.getPTransformRunnerFactories(),
-            IsMapContaining.hasKey(RemoteGrpcPortRead.URN));
-        return;
-      }
+    @Parameterized.Parameter(0)
+    public ProcessBundleSplitResponse expectedResponse;
+
+    @Parameterized.Parameter(1)
+    public long inputElements;
+
+    @Parameterized.Parameter(2)
+    public List<String> processedElements;
+
+    @Parameterized.Parameter(3)
+    public double fractionOfRemainder;
+
+    @Parameterized.Parameter(4)
+    public double elementProgress;
+
+    @Test
+    public void testChannelSplit() throws Exception {
+      SplittingReceiver splittingReceiver = mock(SplittingReceiver.class);
+      BeamFnDataClient mockBeamFnDataClient = mock(BeamFnDataClient.class);
+      when(splittingReceiver.getProgress()).thenReturn(elementProgress);
+      BeamFnDataReadRunner<String> readRunner =
+          createReadRunner(splittingReceiver, PTRANSFORM_ID, mockBeamFnDataClient);
+
+      assertEquals(
+          expectedResponse,
+          executeSplit(
+              readRunner,
+              PTRANSFORM_ID,
+              inputElements,
+              processedElements,
+              fractionOfRemainder,
+              Collections.EMPTY_LIST));
     }
-    fail("Expected registrar not found.");
   }
 
-  @Test
-  public void testSplittingWhenNoElementsProcessed() throws Exception {
-    List<WindowedValue<String>> outputValues = new ArrayList<>();
-    BeamFnDataReadRunner<String> readRunner = createReadRunner(outputValues::add);
-
-    ProcessBundleSplitRequest request =
-        ProcessBundleSplitRequest.newBuilder()
-            .putDesiredSplits(
-                "pTransformId",
-                DesiredSplit.newBuilder()
-                    .setEstimatedInputElements(10)
-                    .setFractionOfRemainder(0.5)
-                    .build())
-            .build();
-    ProcessBundleSplitResponse.Builder responseBuilder = ProcessBundleSplitResponse.newBuilder();
-    readRunner.trySplit(request, responseBuilder);
+  // Test different cases of channel split with non-empty allowed split points.
+  @RunWith(Parameterized.class)
+  public static class ChannelSplitWithAllowedSplitPointsTest {
```

Comment:
The same to (test_split_with_element_allowed_splits)[https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/bundle_processor_test.py#L81]